### PR TITLE
Encoding framework for Haskell-SCENCO

### PR DESCRIPTION
**Pull request by allegroCoder**
Changelog:
- c++ encoding function for generating the opcodes for the partial orders;
- All the algorithm but the SAT-based are implemented;
- Code with unknown and without unknown encoded into the types in Code.hs;
- Several test cases added inside the folder "test";
- Main.hs file includes an example on how to test the encoding function.

### CODE REVIEW

File: Encode.hs [hunk 1]
```diff
@@ -1,7 +1,31 @@
...
 import Code
 import Graph
+import System.IO
+import Foreign
+import Foreign.C.Types
+import Foreign.C.String
+
+testFolder = "test/"
+
+foreign import ccall unsafe "encoding_graphs"
```
Comment (snowleopard) [inline Encode.hs]: Let's follow the same naming convention in all our code: `thisIsAName` instead of `this_is_a_name`.

File: Encode.hs [hunk 2]
```diff
@@ -1,7 +1,31 @@
...
+    encoding_graphs :: CString -> CString -> Int -> IO Int
+
+foreign import ccall unsafe "get_bit"
+    get_bit :: Int -> Int -> IO Int
+
+foreign import ccall unsafe "get_opcodes_length"
+    getOpcodesLength :: Int
+
+data EncodingType = Single_literal |
```
Comment (snowleopard) [inline Encode.hs]: This is usually formatted as follows (I'm also using `camelCase`):
```haskell
data EncodingType = SingleLiteral
                  | Sequential
                  | RandomEncoding
                  | Heuristic
                  | Exhaustive
                  deriving Enum
```

I've also dropped `SatBased` as it is not implemented yet, and renamed `RandomEncoding` to `Random` for consistency.

File: Encode.hs [hunk 3]
```diff
@@ -16,8 +40,66 @@ import Graph
 -- 1) Input codes are conflicting
 -- 2) Input codes are not feasible (not enough bits)
 -- 3) Internal error: gurantees are not satisfied
-encode :: [(Graph, CodeWithUnknowns)] -> Either String [CodeWithoutUnknowns]
-encode = undefined
```
Comment (snowleopard) [inline Encode.hs]: Let's keep these functions commented out in the code for further reference?

File: Encode.hs (hunk 2 above)
Comment (snowleopard) [inline Encode.hs]: Also, don't derive `Enum`. I don't think it is conceptually useful here, but it certainly makes it easier to make a mistake by relying on the order of constructors. If anyone changes the order (or drops one constructor like I did), the code will break.

File: Encode.hs [hunk 4]
```diff
@@ -16,8 +40,66 @@ import Graph
...
+    putStrLn "\t 4) Random encoding (supports constraints)"
+    putStrLn "\t 5) Heuristic encoding (supports constraints)"
+    putStrLn "\t 6) Exhaustive encoding (supports constraints)"
+    putStr "Select algorithm to use: "
+    hFlush stdout
+    encodingNumber <- read <$> getLine
+    return (encodingNumber - 1)
+
+-- uses c++ function to encode the partial orders
+encodeGraphs :: String -> String -> Int -> IO Int
```
Comment (snowleopard) [inline Encode.hs]: There is a special type `FilePath` which should be used instead of `String` here for better readability.

File: Main.hs [hunk 5]
```diff
@@ -4,4 +4,29 @@ import Encode
 import Synthesis
 import TechnologyMapping
 
-main = putStrLn "SCENCO v.0.1"
```
Comment (snowleopard) [inline Main.hs]: Always use type declarations, even for `main :: IO ()`. It's a good habit.

File: Main.hs [hunk 6]
```diff
@@ -4,4 +4,29 @@ import Encode
...
+main = do
+    -- encoding test
+    let numberGraphs = 8 --number of graphs to encode
```
Comment (snowleopard) [inline Main.hs]: Can't you extract 8 from the graph?

File: Main.hs [hunk 7]
```diff
@@ -4,4 +4,29 @@ import Encode
...
+    result <- encodeGraphs poFile customOp algorithm -- encode graphs
+    if result == 0 then do
+        putStrLn "Graphs encoded."
+        let opcodeLength = getOpcodesLength -- get opcode length
+        opcodes <- getOpcodes numberGraphs opcodeLength -- get opcodes
+        print opcodes
+    else putStrLn "Encoding failed."
+
+
+    --if result = 0 then get the matrix
```
Comment (snowleopard) [inline Main.hs]: Drop this commented out code.

File: encoding/encoding.cpp [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+	#include "includes_defines.h"
+	#include "types.h"
+	#include "global.h"
+	#include "utilities.cpp"
+	#include "load_scenarios.cpp"
+	#include "heuristic_function.cpp"
+	#include "custom_encoding.cpp"
+	#include "encoding_methods.cpp"
+#else
+	#include "[absolute_path]\includes_defines.h"
```
Comment (snowleopard) [inline encoding/encoding.cpp]: Why is this needed? I never had to use `[absolute_path]` on Windows! Can you give a link to read about this? Looks like an unnecessary complication to me.

File: encoding/encoding.cpp [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+	#include "[absolute_path]\types.h"
+	#include "[absolute_path]\global.h"
+	#include "[absolute_path]\utilities.cpp"
+	#include "[absolute_path]\load_scenarios.cpp"
+	#include "[absolute_path]\heuristic_function.cpp"
+	#include "[absolute_path]\custom_encoding.cpp"
+	#include "[absolute_path]\encoding_methods.cpp"
+#endif
+
+extern "C" int get_bit(int row, int col){
```
Comment (snowleopard) [inline encoding/encoding.cpp]: Use a single `extern "C" { ... }` block for simplicity.

File: encoding/global.h [hunk 10]
```diff
@@ -0,0 +1,70 @@
+//TEMPORARY FILES
```
Comment (snowleopard) [inline encoding/global.h]: Oh well... :-)

File: encoding/includes_defines.h [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+#include <cstdio>
+#include <cstdlib>
+#include <cstring>
+#include <string>
+#include <vector>
+#include <math.h>
+#include <sstream>
+#include <map>
+
+// constant definition
```
Comment (snowleopard) [inline encoding/includes_defines.h]: Looking forward to getting rid of all these nasty magic numbers.

**Comment by snowleopard** [comment]
Thanks @allegroCoder!

I added a bunch of comments, please do a revision and let me know when complete.

I'm not sure how you are using tests at the moment. Can you try to automate tests in the same way it is done, for example, in the `concepts` project? See https://github.com/tuura/concepts. Maybe @mvdan can help you with this.

File: Encode.hs (hunk 1 above)
Comment (allegroCoder) [inline Encode.hs]: Done! I will be using "thisIsAFunction" naming convention then :+1:

File: Main.hs (hunk 6 above)
Comment (allegroCoder) [inline Main.hs]: I could parse the cpog file for getting the number of POs. I don't think is needed though, since the number of graphs should be defined by the parsing of a Farfalle specification file.

I defined `numberGraphs` for testing purposes only.

File: encoding/encoding.cpp (hunk 8 above)
Comment (allegroCoder) [inline encoding/encoding.cpp]: I used this for compiling the program with codeblocks in Windows. I don't think is needed too. Dropped!

File: encoding/encoding.cpp (hunk 9 above)
Comment (allegroCoder) [inline encoding/encoding.cpp]: DONE!

File: encoding/global.h (hunk 10 above)
Comment (allegroCoder) [inline encoding/global.h]: Do you want me to reduce the amount of global variables? :-)

File: Encode.hs (hunk 2 above)
Comment (allegroCoder) [inline Encode.hs]: I instantiated a function for each encoding at the c++ side of the code. 

```haskell
foreign import ccall unsafe "single_literal_encoding"
    singleLiteralEncoding :: IO Int

foreign import ccall unsafe "sequential_encoding"
    sequentialEncoding :: IO Int

foreign import ccall unsafe "random_encoding"
    randomEncoding :: IO Int

foreign import ccall unsafe "heuristic_encoding"
    heuristicEncoding :: IO Int

foreign import ccall unsafe "exhaustive_encoding"
    exhaustiveEncoding :: IO Int

encodeGraphs :: EncodingType -> IO Int
encodeGraphs e =
    case e of
        SingleLiteral -> singleLiteralEncoding
        Sequential    -> sequentialEncoding
        Random        -> randomEncoding
        Heuristic     -> heuristicEncoding
        Exhaustive    -> exhaustiveEncoding
```

Below is how I convert the number the user introduces for selecting the algorithm to use for encoding:

```haskell
convertAlgorithm :: Int -> EncodingType
convertAlgorithm x
    | x == 0     = SingleLiteral
    | x == 1     = Sequential
    | x == 2     = Random
    | x == 3     = Heuristic
    | x == 4     = Exhaustive
    | otherwise  = error $ "Wrong algorithm selected " ++ show x
```

File: encoding/includes_defines.h (hunk 11 above)
Comment (allegroCoder) [inline encoding/includes_defines.h]: I got rid of the magic numbers I instantiated except for:
-`MAX_NAME`: maximum length of the filename
-`EXHAUSTIVE_MAX_PERM`: maximum number of permutations to try in case exhaustive search is selected

I am not sure how to get rid of the other ones yet though.

Comment (snowleopard) [inline encoding/includes_defines.h]: Why don't you use C++ strings for storing filenames? Then you don't need to specify `MAX_NAME`. `EXHAUSTIVE_MAX_PERM` could instead be passed as a parameter to the heuristic encoding algorithm. Let the user decide how many permutations to try.
